src/relations/postgresql_provider.py

```diff
@@ -109,10 +122,24 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
         if (
             "cluster_initialised" not in self.charm._peers.data[self.charm.app]
             or not self.charm._patroni.member_started
+            or not self.charm.primary_endpoint
         ):
             event.defer()
```

Comment:
I assume you defer since you're waiting for one of these conditions to become true:
```
"cluster_initialised" not in self.charm._peers.data[self.charm.app]
or not self.charm._patroni.member_started
or not self.charm.primary_endpoint
```
In order for some of these to become true some other events need to run (ie the start hook) deferring here can cause this event to continuously defer making it take much longer for the start hook to get on the queue before this event re-queues. This can lead to downtime/poor performance. Instead I suggest you set a flag and trigger and event that will re-call `on_relation_broken`.

Mykola did something similar in his k8s charm [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/lib/charms/mongodb_libs/v0/mongodb_provider.py#L60-L63) and instead of re-triggering the event, he creating a function for himself to call once his condition was met by the start hook, see [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/src/charm.py#L142)

Reply:
Great suggestion Mia! I refactored the code a little bit in https://github.com/canonical/postgresql-operator/pull/17/commits/8bda284c7e8c7bc3d6631ed5aa3b7fc2858989c7 to initially move the user deletion logic to something similar to what you commented. In a future PR I will need to also move the user creation and update.